heron/instance/src/java/org/apache/heron/instance/IInstance.java

```diff
@@ -57,6 +57,21 @@
    */
   void clean();
 
+  /**
+   * Inform the Instance that the framework will clean, stop, and delete the instance
+   * in order to restore its state to a previously-saved checkpoint.
+   *
+   * @param checkpointId the ID of the checkpoint the instance will be restoring to
+   */
+  void preRestore(String checkpointId);
+
+  /**
+   * Inform the Instance that a particular checkpoint has become globally consistent
+   *
+   * @param checkpointId the ID of the checkpoint that became globally consistent
+   */
+  void onCheckpointSaved(String checkpointId);
```

Comment:
These two new interfaces are stateful topology specific. Instead of adding them here in the general `IInstance`, could we find a cleaner way to put them somewhere else?

Reply:
Good point 👍 I will think about it

Reply:
On the other side, init() and persistState() are in this interface, so I feel it might be ok.